### ISSUE TITLE
Native Federation Supporting Nested Modules

### ISSUE THREAD

**Issue by FlorianRappl** [angular-architects/module-federation-plugin#284]
Right now, the Native Federation (NF) plugin has an issue with nested modules. Using Webpack Module Federation (WMF) what you can do is to have the following picture:

```
# host
  -> remotes: MF1
  -> exposes: -
  -> entry: index.ts with initFederation (necessary to teach about MF1)
# MF1
  -> remotes: MF2, MF3
  -> exposes: Some Page (Page.tsx)
  -> entry: index.ts with initFederation (necessary to teach about MF2, MF3)
# MF2
  -> remotes: -
  -> exposes: Component 1 (Component1.tsx), Component 2 (Component2.tsx)
  -> entry: index.ts with initFederation (not used /  necessary)
# MF3
  -> remotes: -
  -> exposes: Component 3 (Component3.tsx)
  -> entry: index.ts with initFederation (not used / necessary)
```

So while host and MF2/3 are truly fulfilling their roles, MF1 is kind of a hybrid. So far, so good. In WMF this is all working as the remotes are declared in the same manifest, however, in NF this is split up. The module gets to know the remotes via the `initFederation` function. But `initFederation` would only be run on the host, and exposed modules should not include it (at least, this is how the examples work and how, in general, it makes sense).

Now, there are multiple ways out of this scenario. Let's start simple: If the information registered / known from the host / web app could be read by the MFs we could just change the host to

```
# host
  -> remotes: MF1, MF2, MF3
  -> exposes: -
  -> entry: index.ts with initFederation (necessary to teach about MF1 and children like MF2 and MF3)
```

However, right now this is not possible as the `@softarc/native-federation` package is always excluded as shared dependency. Hence, it will be always bundled and the information is never shared.

Of course, even with this bundling we could share it, e.g., by replacing ([native-federation-runtime](https://github.com/angular-architects/module-federation-plugin/tree/main/libs/native-federation-runtime)/[src](https://github.com/angular-architects/module-federation-plugin/tree/main/libs/native-federation-runtime/src)/[lib](https://github.com/angular-architects/module-federation-plugin/tree/main/libs/native-federation-runtime/src/lib)/[model](https://github.com/angular-architects/module-federation-plugin/tree/main/libs/native-federation-runtime/src/lib/model)/remotes.ts)

```ts
const remoteNamesToRemote = new Map<string, Remote>();
const baseUrlToRemoteNames = new Map<string, string>();
```

with

```ts
const remoteNamesToRemote = (window.nfRemotesNamesToRemote ??= new Map<string, Remote>());
const baseUrlToRemoteNames = (window.nfBaseUrlToRemoteNames ??= new Map<string, string>());
```

Hence, the infos about the remotes are shared on the global object. Personally, I don't like this workaround. However, it's still one option.

The thing that I don't like is that the host would need all the information from the remotes to place also the nested remotes. I mean, yes, my / our recommendation is anyway to keep it flat, but since this system allows you to make it nested / deep, you should be able to do it - without having to flatten / grow it out on the host.

That being written an imho nicer way would be to do three things:

1. Share the runtime. With a shared runtime a host could also dynamically set / change what MFs are discovered / are there. The runtime itself could be registered as an esm that can be imported just like `import '@softarc/native-federation-runtime` or so (the importmap entry etc. would set up by the bundled part of the host, but exposed parts / remotes would not need to bundle it or set it up).
2. Already declare the fallback / implicit usage on the other remotes in the manifest (`remoteEntry.json`). This way, the runtime has a default to fallback to when the current host has not overridden it.
3. Add hooks to allow the MF resolution to be overridden / changed. The hooks are just standard DOM custom events.

Any thoughts? Some other way?

**Comment by manfredsteyer** [comment]
Good point. Thanks for pointing this out, @FlorianRappl. We should keep this in mind for one of the next versions.
